channeldb/kvdb/etcd/commit_queue.go

```diff
@@ -0,0 +1,139 @@
+// +build kvdb_etcd
+
+package etcd
+
+import (
+	"context"
+	"sync"
+)
+
+// commitQueueSize is the maximum number of commits we let to queue up. All
+// remaining commits will block on commitQueue.Add().
+const commitQueueSize = 100
+
+// commitQueue is a simple execution queue to manage conflicts for transactions
+// and thereby reduce the number of times conflicting transactions need to be
+// retried. When a new transaction is added to the queue, we first upgrade the
+// read/write counts in the queue's own accounting to decide whether the new
+// transaction has any conflicting dependencies. If the transaction does not
+// conflict with any other, then it is comitted immediately, otherwise it'll be
+// queued up for later exection.
+type commitQueue struct {
+	ctx       context.Context
+	mx        sync.Mutex
+	readerMap map[string]int
+	writerMap map[string]int
+
+	commitMutex sync.RWMutex
+	queue       chan (func())
+	done        chan (struct{})
+}
+
+// NewCommitQueue creates a new commit queue, with the passed abort context.
+func NewCommitQueue(ctx context.Context) *commitQueue {
+	q := &commitQueue{
+		ctx:       ctx,
+		readerMap: make(map[string]int),
+		writerMap: make(map[string]int),
+		queue:     make(chan func(), commitQueueSize),
+		done:      make(chan struct{}),
+	}
+
+	// Start the queue consumer loop.
+	go q.mainLoop()
+
+	return q
+}
+
+// Wait waits for the queue to stop (after the queue context has been canceled).
+func (c *commitQueue) Wait() {
+	<-c.done
+}
+
+// Add increases lock counts and queues up tx commit closure for execution.
+// Transactions that don't have any conflicts are executed immediately by
+// "downgrading" the count mutex to allow concurrency.
+func (c *commitQueue) Add(commitLoop func(), rset readSet, wset writeSet) {
+	c.mx.Lock()
+	blocked := false
+
+	// Increase read counts and mark as blocked if any key is in the
+	// writer map.
+	for key := range rset {
+		c.readerMap[key] += 1
+		if !blocked {
+			blocked = c.writerMap[key] > 0
+		}
+	}
+
+	// Increase write counts and mark as blocked if any key is in the
+	// reader map or is already blocked for writing.
+	for key := range wset {
+		c.writerMap[key] += 1
+		if !blocked {
+			_, rsetContainsKey := rset[key]
+			blocked = (c.writerMap[key] > 1 ||
+				(c.readerMap[key] > 0 && !rsetContainsKey))
```

Comment:
Why don't we need to block if there's a pending transaction in the queue that reads this key, we want to write it, but don't also read the key ourselves?

Reply:
It's because if our read set contains the key then we already increased `c.readerMap[key]` above, so to make sure reader lock count is non zero we have to "uncount ourselves".

Reply:
I think this would be much easier to reason about by just by doing two passes through the sets:
```go
for key := range rset {
    blocked |= c.writerMap[key] > 0
}
for key := range wset {
    blocked |= c.writerMap[key] > 0 || c.readerMap[key] > 0
}
for key := range rset {
    c.readerMap[key] += 1
}
for key := range wset {
    c.writerMap[key] += 1
}
```

Performance wise I doubt we'll see any difference.

Reply:
** and also simplified a bit to make it more? readable